src/reader/decoder.rs

```diff
@@ -754,17 +766,39 @@ impl StreamingDecoder {
                         }
                     }
                 } else {
+                    self.ext.data.push(b);
                     self.ext.is_block_end = false;
                     goto!(ExtensionDataBlock(b as usize), emit Decoded::SubBlockFinished(self.ext.id))
                 }
             }
             ApplicationExtension => {
                 debug_assert_eq!(0, b);
                 // the parser removes sub-block lenghts, so app name and data are concatenated
-                if self.ext.data.len() >= 15 && &self.ext.data[1..13] == b"NETSCAPE2.0\x01" {
-                    let repeat = &self.ext.data[13..15];
+                if self.ext.data.len() >= EXT_NAME_NETSCAPE.len()
+                    && &self.ext.data[..EXT_NAME_NETSCAPE.len()] == EXT_NAME_NETSCAPE
+                {
+                    let repeat = &self.ext.data[EXT_NAME_NETSCAPE.len()..];
                     let repeat = u16::from(repeat[0]) | u16::from(repeat[1]) << 8;
                     goto!(BlockEnd, emit Decoded::Repetitions(if repeat == 0 { Repeat::Infinite } else { Repeat::Finite(repeat) }))
+                } else if self.ext.data.len() >= EXT_NAME_XMP.len()
+                    && &self.ext.data[..EXT_NAME_XMP.len()] == EXT_NAME_XMP
+                {
+                    self.ext.data.drain(..EXT_NAME_XMP.len());
+                    // XMP adds a "ramp" of 257 bytes to the end of the metadata to let the "pascal-strings"
+                    // parser converge to the null byte. The ramp looks like "0x01, 0xff, .., 0x01, 0x00".
+                    // For convenience and to allow consumers to not be bothered with this implementation detail,
+                    // we cut the ramp.
+                    const RAMP_SIZE: usize = 257;
+                    if self.ext.data.len() >= RAMP_SIZE
+                        && self.ext.data.ends_with(&[0x03, 0x02, 0x01, 0x00])
+                        && self.ext.data[self.ext.data.len() - RAMP_SIZE..]
+                            .starts_with(&[0x01, 0x0ff])
+                    {
+                        self.ext.data.truncate(self.ext.data.len() - RAMP_SIZE);
+                    }
+                    // We can consume the extension data here as the next states won't access it anymore.
+                    self.xmp_metadata = Some(std::mem::take(&mut self.ext.data));
```

Comment:
Nothing is going to need the data later, so maybe do `data = mem::take` at the beginning and then use local var?

Reply:
Fair!